### PR TITLE
Optimize line item merging

### PR DESCRIPTION
**Pull request by dkubb**

### CODE REVIEW

File: core/app/models/spree/line_item.rb [hunk 1]
```diff
@@ -35,11 +33,12 @@ def adjustments
       order.all_adjustments.adjustable(self)
     end
 
-    def copy_price
-      if variant
-        self.price    = variant.price    if price.nil?
-        self.currency = variant.currency if currency.nil?
-      end
+    def price
+      self[:price] ||= variant&.price
```
Comment (mbj) [inline core/app/models/spree/line_item.rb]: Any reason we cannot use `self.price ||= variant&.price` here?

Comment (dkubb) [inline core/app/models/spree/line_item.rb]: Wouldn't that cause a recursion issue?

Comment (mbj) [inline core/app/models/spree/line_item.rb]: > Wouldn't that cause a recursion issue? 

AFAIK not, the RHS of the op-asgn is only peeking into the variant, and the variant will not have the line item in scope to trigger a recursion.

Comment (dkubb) [inline core/app/models/spree/line_item.rb]: I am going to make sure this method is fully mutation covered then I'll give this a test.

File: core/app/models/spree/line_item.rb [hunk 2]
```diff
@@ -35,11 +33,18 @@ def adjustments
...
+    # @return [BigDecimal]
+    def price
+      self[:price] ||= variant&.price
+    end
+
+    # The line item currency
+    #
+    # @return [String]
+    def currency
+      self[:currency] ||= variant&.currency
```
Comment (mbj) [inline core/app/models/spree/line_item.rb]: Should we simply forward `LineItem#currency` to `Order#curency`? Or is this too big of a change?

Comment (dkubb) [inline core/app/models/spree/line_item.rb]: Eventually yes, but I'd like to do that in another pass later. I'm trying to do the minimum possible to optimize line item merging atm.

Comment (mbj) [inline core/app/models/spree/line_item.rb]: > Eventually yes, but I'd like to do that in another pass later. I'm trying to do the minimum possible to optimize line item merging atm.

Agreed, just to notice this optimizations will have an effect beyond line item merging, as all operations handling promotions will be positively affected also.

File: core/spec/models/spree/line_item_spec.rb [hunk 3]
```diff
@@ -23,15 +26,85 @@
...
+      subject.currency
+    end
+
+    shared_examples '#currency' do |expected|
+      it 'returns the expected currency' do
+        expect(apply).to eql(expected)
+      end
+
+      it 'is idempotent' do
+        expect(apply).to be(apply)
```
Comment (mbj) [inline core/spec/models/spree/line_item_spec.rb]: AFAIK we do have a dedicated `is_idempoent` helper for this already. `is_idempotent { apply }` should do the job here.

Comment (dkubb) [inline core/spec/models/spree/line_item_spec.rb]: This helper is not available in this repo, only the mrh-spree one.

**Pull request merged**